### PR TITLE
Switching to boto3 only

### PR DESCRIPTION
**Pull request by revmischa**
I have removed all traces of boto2 that I could find. I got async message signing and parsing more or less working. Probably still needs some cleanup and testing. 
Much farther along now on the journey to boto3.

### CODE REVIEW

File: kombu/transport/SQS.py [hunk 1]
```diff
@@ -160,15 +152,23 @@ def _reset_cycle(self):
 
     def entity_name(self, name, table=CHARS_REPLACE_TABLE):
         """Format AMQP queue name into a legal SQS queue name."""
-        return text_t(safe_str(name)).translate(table)
```
Comment (georgepsarakis) [inline kombu/transport/SQS.py]: @revmischa why is there a differentiation for queues with suffix `.fifo`?

Comment (revmischa) [inline kombu/transport/SQS.py]: @jseutter did that not me. 
.fifo suffix means it's a FIFO queue, i.e. retains ordering. this is a new feature in SQS. non-FIFO queues don't preserve ordering.

File: requirements/extras/sqs.txt [hunk 2]
```diff
@@ -1,2 +1,3 @@
-boto>=2.8
+boto3
+botocore
```
Comment (georgepsarakis) [inline requirements/extras/sqs.txt]: I think `boto3` already depends on `botocore`. Also, perhaps it would be a good idea to specify (at least) a minimum version?

File: t/unit/async/aws/case.py [hunk 3]
```diff
@@ -8,6 +8,7 @@
 
 @skip.if_pypy()
 @skip.unless_module('boto')
```
Comment (georgepsarakis) [inline t/unit/async/aws/case.py]: Perhaps this line should be removed?

File: kombu/transport/SQS.py (hunk 1 above)
Comment (georgepsarakis) [inline kombu/transport/SQS.py]: Is that a necessary change in the scope of removing boto2? I am pointing this out because the changes are already very extensive. I think it would be advisable to keep the diff to a minimum.

Comment (revmischa) [inline kombu/transport/SQS.py]: I think FIFO support is absolutely a requirement. Otherwise Kombu will likely break if someone tries to use a new-style queue. However I did not implement this, I'm including the changes from @jseutter in my PR because they're based off of his branch.

File: requirements/extras/sqs.txt (hunk 2 above)
Comment (revmischa) [inline requirements/extras/sqs.txt]: Yes, good point
https://github.com/celery/kombu/pull/693/commits/17f4ab00c3ad9575a8c97e183b841a44dd00c37a

File: t/unit/async/aws/case.py (hunk 3 above)
Comment (revmischa) [inline t/unit/async/aws/case.py]: yup

File: kombu/transport/SQS.py (hunk 1 above)
Comment (georgepsarakis) [inline kombu/transport/SQS.py]: So, it is basically an enhancement, adding SQS FIFO support :) . I believe it belongs to a separate Pull Request, but I will not insist any further.

**Comment by georgepsarakis** [comment]
I found this [project](https://github.com/adamw/elasticmq#using-the-amazon-boto-python-to-access-an-elasticmq-server) that seems quite active. Perhaps it could be used for integration tests? What do you think? Could be useful for Celery testing as well.

File: kombu/transport/SQS.py (hunk 1 above)
Comment (revmischa) [inline kombu/transport/SQS.py]: I don't disagree with that. It's not my commit though.

**Comment by revmischa** [comment]
Maybe usesful @georgepsarakis I'm just trying to get all the wacky original SQS tests passing again which is no small feat. They were extremely coupled to the boto2 behaviors.

**Comment by auvipy** [comment]
it's better to squash commits before pushing. this creates unnecessary noise to build system

**Comment by revmischa** [comment]
Sorry, I don't know how to run the multiple python environment tests locally or landscape locally. So I have to push and wait for CI.

**Comment by jseutter** [comment]
@revmischa - Are you interested in working together on Kombu's SQS support?  I've been thinking about splitting up my original PR into smaller ones:
- Changes would be easier for the maintainers to digest, have a better chance of getting merged.
- We could collaborate on removing the boto 2 dependency

Thoughts?

**Comment by revmischa** [comment]
@jseutter check out the work I did and let me know if you have any questions. i'm not planning to spend any more time on this right this second. theres no boto2 left anywhere.

**Comment by auvipy** [comment]
@revmischa thanks for the work. what reamain in your thought?

@jseutter review and comment what need to be change

**Comment by revmischa** [comment]
I managed to get all the tests passing, but I wasn't sure if all the modifications I made were proper. Really the code needs review by someone who understands Kombu better than me. I was trying to perform some major surgery without having any idea of how anything was supposed to fit together. I changed everything I could find but I wasn't really sure how to test all the cases. Probably the channel creation (esp with FIFO support) needs proper testing. I also made big changes to the message serialization/deserialization stuff. 
tl;dr I don't know what's left to do. I know zero about Kombu and probably broke some things.

**Comment by georgepsarakis** [comment]
@revmischa could you please rebase?

**Comment by revmischa** [comment]
You rebase it. I gave up on Celery and wrote something better using PostgreSQL's `SELECT ... FOR UPDATE SKIP LOCKED` feature.

**Comment by alukach** [comment]
@revmischa So are you (and the rest of the @jetbridge organization) formally abandoning this PR?

I'm interested in getting this merged. Getting away from Boto will be a large win for Kombu moving forward as boto2:

* doesn't support Python3.4 (boto/boto#3713)
* doesn't support FIFO queues (celery/kombu#678, celery/celery#3690)
* has broken endpoints files (celery/celery#3672)
* appears to be dead/overwhelmed (at time of writing, there are 359 open PRs and no activity for the past month)

@georgepsarakis What are your feelings about this PR? How far is it from completion? What else needs to be done before you'd feel comfortable merging? I'm just stepping into this code right now but am willing to put in some hours if need be. I'm running off of the PR branch on my local machine and it seems to be close to working (my message bodies appear to be encoded in Base64, need to figure out why that is, quite possibly not related to this PR).

@jseutter Do you have any thoughts on what needs to be done for this to be called "complete"?

**Comment by revmischa** [comment]
I'll merge in changes if you want but I don't have anything further to contribute to this PR personally.

**Comment by auvipy** [comment]
a rebase would be fine

**Comment by auvipy** [comment]
i rebased and squashed. plz test and report bugs

**Comment by georgepsarakis** [comment]
@revmischa thanks for your efforts on this PR :smiley: .

@alukach @auvipy I will try to add integration tests in Celery using [Atlassian LocalStack](https://github.com/atlassian/localstack), which I recently discovered, and has SQS support.

**Comment by auvipy** [comment]
that would be really awesome!

File: kombu/transport/SQS.py [hunk 4]
```diff
@@ -191,15 +194,27 @@ def _delete(self, queue, *args, **kwargs):
...
+            if 'MessageDeduplicationId' in message['properties']:
+                kwargs['MessageDeduplicationId'] = \
+                    message['properties']['MessageDeduplicationId']
+            else:
+                kwargs['MessageDeduplicationId'] = str(uuid.uuid4())
+        self.sqs.send_message(**kwargs)
 
     def _message_to_python(self, message, queue_name, queue):
-        payload = loads(bytes_to_str(message.get_body()))
+        payload = loads(bytes_to_str(message['Body']))
```
Comment (alukach) [inline kombu/transport/SQS.py]: At this moment, won't the message be in in `base64`?

Recommend changing this to:
```python
body = base64.b64decode(message['Body'].encode())
payload = loads(bytes_to_str(body))
```

File: kombu/transport/SQS.py [hunk 5]
```diff
@@ -209,14 +224,14 @@ def _message_to_python(self, message, queue_name, queue):
                 delivery_info = {}
                 properties = {'delivery_info': delivery_info}
                 payload.update({
-                    'body': bytes_to_str(message.get_body()),
+                    'body': bytes_to_str(message['Body']),
```
Comment (alukach) [inline kombu/transport/SQS.py]: ... and then use the `body` variable here.

File: kombu/transport/SQS.py (hunk 4 above)
Comment (alukach) [inline kombu/transport/SQS.py]: @revmischa I know you're signing off on this PR, however would you be able to comment on this?  I was unable to get the code running with modifying the code as mentioned above.

**Comment by alukach** [comment]
@auvipy @georgepsarakis Thanks for rebasing, however I think the merge may have been a bit pre-emptive.  Did anyone try to test this on their machines?  I was unable to get workers to consume tasks without modifying the code.

File: kombu/transport/SQS.py (hunk 4 above)
Comment (revmischa) [inline kombu/transport/SQS.py]: I am not sure why base64 was used, if I recall correctly it was done in the boto2 version and I didn't change it. I was developing on python3.5 but I definitely had some issues with the encoding/decoding differences between py2 and p3 and I suggest anything you can do to simplify all of that is a good idea.

File: kombu/async/aws/sqs/message.py [hunk 6]
```diff
@@ -2,45 +2,32 @@
...
-
-class AsyncEncodedMHMessage(BaseAsyncMessage, EncodedMHMessage):
-    """Encoded MH Message."""
-
-
-class AsyncJSONMessage(BaseAsyncMessage, JSONMessage):
-    """Json serialized message."""
+    def encode(self, value):
+        """Encode/decode the value using Base64 encoding."""
+        return base64.b64encode(value).decode('utf-8')
```
Comment (alukach) [inline kombu/async/aws/sqs/message.py]: I've since learned that encoding to Base64 is recommend at it appears that Boto3 does not do this automatically. Deleting the above comment.

File: kombu/transport/SQS.py (hunk 4 above)
Comment (georgepsarakis) [inline kombu/transport/SQS.py]: @alukach you can see that the base64 encoding is related to the Message implementation of boto [here](https://github.com/boto/boto/blob/master/boto/sqs/message.py#L154).

**Comment by georgepsarakis** [comment]
@alukach would you mind opening a separate Pull Request with your changes? It would be valuable to merge fixes since you seem to be testing with SQS directly. Thanks!

File: kombu/transport/SQS.py (hunk 4 above)
Comment (alukach) [inline kombu/transport/SQS.py]: Yup, I now understand that the boto2 did the base64 encoding automatically. Will make a PR in a few...

**Comment by alukach** [comment]
@georgepsarakis PR made. Take a look. I'll be signing off for the evening but will be back and available for comments tomorrow.

**Comment by audiolion** [comment]
Yay! 🎉 

@revmischa Thank you thank you for all the initial work done to support SQS!! it is much appreciated

Thank you @alukach and kombu team for iterating on this patch, this was holding me back from using celery + sqs!

**Comment by krzysztofwolski** [comment]
Thanks! I was waiting for this merge :). When can I expect release with those changes?
